docs/logicapps-ai-course/02_build_conversational_agents/06-extend-tools-with-patterns.md

```diff
@@ -8,10 +8,74 @@ author: karansin
 ms.author: karansin
 ---
 
-Tool patterns - Mocking, Guardrails, Output transformation
+# Module 06 - Extend your tool functionality with patterns
 
-- Using compose action for mocking
-- Using multiple actions in a tool branch
-- How to implement Guardrails using the above
-- Using compose action to transform the output of a tool
-- This is grabbag of patterns - will write content and we can move it later if doesn't fit
+This module describes several common tool patterns:
+- Mocking tool output
+- Adding multiple actions per tool
+- Transforming tool output
+- Complex control flow in tools via nested workflows
+- Parallel tool execution
+- Human-in-the-loop
+
+## Mocking tool output
+
+When developing your agent, it can be helpful to iterate on high-level agent design (model selection, system prompt, tool metadata) without real tool execution. For example, you may want to defer connector configuration or avoid side effects.
+
+To accomplish this, you can mock tool output in two ways:
+- Use a Compose action
+- Use static results
+
+### Mocking using Compose action
+
+1. When implementing the tool, instead of configuring your intended connector, select "Compose" action. This is a built-in action in Logic Apps that allows you to compose static or dynamic values into a final result.
+2. Replace the Compose action Inputs with the value you would like the tool to return.
+3. When executing the tool, the Compose action will return your provided value as-is.
+
+See the `EchoTool` example in module four.
+
+### Mocking using static results
+
+Consult [this documentation](https://learn.microsoft.com/en-us/azure/logic-apps/testing-framework/test-logic-apps-mock-data-static-results?tabs=standard) to set up static results on the action whose execution you would like to mock.
+
+## Adding multiple actions per tool branch
+
+So far, all our examples have had one Logic App action per agent tool branch. You also have the option to add multiple linear actions. To do this, simply select the "Add" button within a tool to set up multiple actions.
+
+When a tool completes execution, the results must be sent back to the LLM for interpretation. In Logic Apps, the result of the **final** action in your tool branch becomes the tool output.
+
+## Transforming tool output
+
+The final action in your tool branch may return an intermediate payload that you want to transform further before sending back to LLM for interpretation.
+
+This capability has several use cases:
+- Reduce overall token count by only sending minimal information to LLM
+- Improve agent quality by filtering out unnecessary information from the tool output
+- By default, the last action output becomes the tool output. If you have a chain of several actions, you can construct a custom tool output payload that combines properties from multiple action outputs. 
+
+This can again be accomplished with the Compose action.
```

Comment:
I will add a proper example for this one in separate PR since it gets into expressions